File: Chapter02/python/packt2/spark/warc_schema_parser.py
from utilities01_py.helper_python import create_session
from utilities02_py.helper_python import extract_raw_records, parse_raw_warc
from globalp.python.packtg.helper_python_global import sample_warc_loc

if __name__ == "__main__":
    warc_loc = sample_warc_loc
    session = create_session(3, 'Schema Parsing WARC')
    session.sparkContext.setLogLevel('ERROR')  # avoids printing of info messages

    raw_records = extract_raw_records(warc_loc, session)
    warc_records = raw_records \
        .flatMap(lambda record: parse_raw_warc(record))

    warc_records.toDF().printSchema()
    print('Total records: ' + str(warc_records.count()))

    # responses = warc_records \
    #     .filter(lambda record: record.warc_type == "response") \
    #     .toDF()
    #
    # responses.printSchema()
    # responses.show(3)
    #
    # english_records = responses.filter(responses.language == 'en')
    # print(english_records.count())
    # # 1
    # print(english_records.select('html_source').first())

File: Chapter02/scala/packt2/spark/WarcSchemaParser.scala
//package packt2.spark
//
//import org.apache.hadoop.io.Text
//import org.apache.spark.rdd.RDD
//import org.apache.spark.sql.SparkSession
//import Utilities01.HelperScala.createSession
//import Utilities02.HelperScala.{sampleWarcLoc, extractRawRecords, parseRawWarc}
//
///**
// * Code for parsing .warc files of the WARC corpus
// *
// * @author Phil, https://github.com/g1thubhub
// */
//object WarcSchemaParser {
//
//  def main(args: Array[String]) = {
//
//    val inputLocWarc = sampleWarcLoc
//    implicit val session: SparkSession = createSession(3, "WARC Parser")
//    session.sparkContext.setLogLevel("ERROR")  // avoids printing of info messages
//
//    val rawRecords: RDD[Text] = extractRawRecords(inputLocWarc)
//    val warcRecords: RDD[WarcRecord] = rawRecords
//      .flatMap(parseRawWarc(_))
//  CURRENT TIME PRINT OUT
//    import session.implicits._
//    warcRecords.toDF().printSchema()
//    println(s"Total records: ${warcRecords.count()}")
//
//    /*
//    import org.apache.spark.sql._
//    val responses: DataFrame = warcRecords
//      .filter(_.warcType == "response")
//      .toDF()
//    responses.printSchema()
//
//    responses.show(3)
//
//    // small subset of overall English records which were explicitly marked in server response, see scratchpad.txt
//    val englishRecords = responses.filter($"language" === "en")
//    println(englishRecords.count())
//    // 1
//    println(englishRecords.map(_.getAs[String]("htmlSource")).first())
//  */
//  }
//}
